src/inmanta/ast/statements/generator.py

```diff
@@ -268,7 +267,151 @@ def nested_blocks(self) -> Iterator["BasicBlock"]:
         yield self.module
 
 
-class If(ExpressionStatement):
+# TODO: include general flow in docstring (here or requires_emit?): requires_emit -> resume -> execute
+class ListComprehension(RawResumer, ExpressionStatement):
+    """
+    A list comprehension expression, e.g. `["hello {{world}}" for world in worlds]`.
+    """
+
+    __slots__ = ("loop_var", "value_expression", "iterable")
+
+    def __init__(self, value_expression: ExpressionStatement, loop_var: LocatableString, iterable: ExpressionStatement) -> None:
+        super().__init__()
+        self.value_expression: ExpressionStatement = value_expression
+        self.loop_var: LocatableString = loop_var
+        self.iterable: ExpressionStatement = iterable
+        # TODO: extend anchors?
+
+    def normalize(self, *, lhs_attribute: Optional[AttributeAssignmentLHS] = None) -> None:
+        self.value_expression.normalize(lhs_attribute=lhs_attribute)
+        self.iterable.normalize()
+
+    def requires(self) -> list[str]:
+        # exclude loop var, unless it shadows an occurrence in iterable
+        return list(set(self.iterable.requires()) - {self.loop_var} + set(self.value_expression.requires()))
+
+    def requires_emit(
+        self, resolver: Resolver, queue: QueueScheduler, *, lhs: Optional[ResultCollector[object]] = None
+    ) -> dict[object, VariableABC]:
+        """
+        Sets up gradual execution of the list comprehension. Additionally sets up a resumer for when the iterable is complete.
+        Returns as requires the gradual helper's result variable, which will be frozen by the resumer.
+
+        Flow: requires_emit -> gradual execution -> resume -> execute
+        """
+        base_requires: dict[object, VariableABC] = super().requires_emit(resolver, queue)
+
+        # set up gradual execution
+        gradual_helper: ListComprehensionGradual = ListComprehensionGradual(
+            statement=self,
+            resolver=resolver,
+            queue=queue,
+            lhs=lhs,
+        )
+        iterable_requires: dict[object, VariableABC] = self.iterable.requires_emit_gradual(
+            resolver, queue, resultcollector=gradual_helper
+        )
+
+        # non-gradual mode / finishing up
+        # pass helper to the resumer via the requires object
+        wrapped_helper: VariableABC = WrappedValueVariable(gradual_helper)
+        requires: dict[object, VariableABC] = base_requires | iterable_requires | {self: wrapped_helper}
+        RawUnit(queue, resolver, requires, resumer=self)
```

Comment:
Would it not be easier to make the `ListComprehensionGradual`  the resumer as well to get rid of the WrappedValueVariable?

Reply:
I rejected that for some reason. I'll look it over once more and get back to you.

Reply:
It is already a resumer of itself now so I think the current implementation is the best we can get.